Reservoir coupling: Communicate group potentials from slaves and compute guide rates

hakonhagland opened the pull request:
Depends on https://github.com/OPM/opm-common/pull/4567

Sends group potentials from the slave processes to the master process, then use those potentials in the master process to calculate guide rates.

A future PR will calculate group target in the master based on the guide rates and send those targets to the slave groups.

hakonhagland commented:
I will run jenkins from the opm-common PR later

hakonhagland commented on opm/simulators/flow/ReservoirCouplingSlave.cpp:
```diff
@@ -268,9 +300,64 @@ saveMasterGroupNamesAsMap_(const std::vector<char>& group_names) {
+{
+    if (this->comm_.rank() == 0) {
+        // NOTE: The master process will use this date to check that no slave starts before the master
+        double activation_date = this->getGrupSlavActivationDate_();
```
This was intentionally named GrupSlav (and not GrupSlave) because the keyword name is `GRUPSLAV`.

hakonhagland commented on opm/simulators/flow/ReservoirCouplingSlave.cpp:
```diff
@@ -256,6 +209,85 @@ maybeActivate(int report_step) {
+    }
+    this->comm_.broadcast(group_names.data(), /*count=*/size, /*emitter_rank=*/0);
+    this->saveMasterGroupNamesAsMap_(group_names);
+    this->checkGrupSlavGroupNames_();
```
Same reason as for the previous comment.

blattms reviewed (commented):
Thanks a lot for the contribution. A few comments for now.

Note that this does not build without MPI.

blattms commented:
Unfortunately, this needs to be rebased on master it seems.

hakonhagland commented:
@blattms Thanks for the review with useful suggestions. I have added some commits to the PR to address the comments. I also added one commit to dump all guide rates to the debug file.

blattms reviewed (commented):
Thanks  a lot for rebasing.
This is looks quite good already.

My main reservation is the introduction of a reference to the simulator in the time stepping. I would hope that there is another way.

blattms commented:
jenkins build this serial please

blattms commented:
:see_no_evil: jenkins will fail of course. Handled by opm-common PR, already

hakonhagland commented on opm/simulators/timestepping/AdaptiveTimeStepping.hpp:
```diff
@@ -125,6 +127,7 @@ class AdaptiveTimeStepping
         Solver& solver_;
         const bool is_event_;
         const TuningUpdateCallback& tuning_updater_;
+        Simulator& simulator_;
```
I did not think about that. Since the user can anyway abuse simulator by going through `solver.model().simulator()` and we need the solver reference anyway, storing the simulator is just a convenience for writing `this->simulator_` instead of `this->solver_.model().simulator()`, right? See, the added commit.

hakonhagland commented:
> My main reservation is the introduction of a reference to the simulator in the time stepping.

@blattms Thanks for the detailed review. I have added some commits to address the comments you posed. I am also planning to add a commit to address [the comment](https://github.com/OPM/opm-common/pull/4567#issuecomment-2963363992) from @bska in the companion PR in opm-common. I think we can avoid getting the slave group potentials from the `GuideRate` object by storing them in the `GroupState` object instead. This will for now cause duplication of group potentials in `GuideRate` and `GroupState`. A future PR might try to remove that duplication by removing the potentials from the `GuideRate` object's internal state. What do you think? Then we can simply close/delete https://github.com/OPM/opm-common/pull/4567.

hakonhagland commented:
jenkins build this please

hakonhagland commented:
jenkins build this please

hakonhagland commented:
jenkins build this please

hakonhagland commented:
> The last commit tries to address OPM/opm-common#4567 (comment). Then OPM/opm-common#4629 will replace OPM/opm-common#4567.

@blattms and @bska After review by @bska, I think the companion PR https://github.com/OPM/opm-common/pull/4629 can also be eliminated. I have added a [commit](https://github.com/OPM/opm-simulators/pull/6170/commits/9823ebc6ffe52795afc1cf79b970c359e1a2dde9) here as a replacement, and plan to close both opm-common PRs as they are no longer needed. Jenkins builds with and without serial succeed. Let me know if anything else needs to be added to this PR.

blattms commented:
Thanks a lot. I'll take another look.

blattms commented on opm/simulators/timestepping/AdaptiveTimeStepping.hpp:
```diff
@@ -78,6 +78,8 @@ class AdaptiveTimeStepping
 
 private:
     using Scalar = GetPropType<TypeTag, Properties::Scalar>;
+    using Simulator = GetPropType<TypeTag, Properties::Simulator>;
```
Is this used?

blattms commented on opm/simulators/flow/ReservoirCouplingMpiTraits.hpp:
```diff
@@ -0,0 +1,154 @@
+        MPI_Get_address(&(d.*Member), &disp[idx]);
+        disp[idx] -= base;
+
+        if constexpr (std::is_array_v<MemberT>) {
+            // C array  T[N]
+            using Elem = std::remove_extent_t<MemberT>;
+            blk  [idx] = std::extent_v<MemberT>;
+            types[idx] = MPITraits<Elem>::getType();
+        }
```
This should also work if T is not an array and give the same results as the else branch (you still need to handle num). Can we simplify this?

What branches are actually used? This looks quite general, but the surrouding class is not.

blattms reviewed (commented):
Overall this looks good. Just a few curious questions.

Personally, I would appreciate a bit more doxygen documentation for new classes.
